### ISSUE TITLE
[CRITICAL] [UX Reliability] App Crash with Invalid Currency code

### ISSUE THREAD

**Issue by mountiny** [Expensify/App#43004]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:**
**Reproducible in staging?:**
**Reproducible in production?:**
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @joekaufmanexpensify 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1717433997016439

## Action Performed:
Break down in numbered steps

1. Have a report with invalid currency code
2. You can locally merge in incorrect data to an expense report
3. Open the report

## Expected Result:
Describe what you think should've happened

We will handle this issue gracefully

## Actual Result:
Describe what actually happened

The app crashes 
![image](https://github.com/Expensify/App/assets/36083550/14ff25f3-1073-4093-ab99-55dfc857d96d)

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by mountiny** [comment]
Looking for proposals

**Comment by b4s36t4** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
[UX Reliability] App Crash with Invalid Currency code

### What is the root cause of that problem?


`Intl.NumberFormat` does have an issue when we either pass empty currency code or currency code which is greater than 4 characters.

If we pass an invalid key say `OOO` it still works, if I pass `oooo` or empty valye the application breaks.

So the real issue relies on sending empty currency code which is happening here https://github.com/Expensify/App/blob/926d7af225d72feb3b8a958a1fc97fa99ee563fb/src/libs/ReportUtils.ts#L6372-L6386

If we observe here we're passing empty string here which can cause application break.

### What changes do you think we should make in order to solve the problem?
So, to solve this we have to ensure we never pass empty currency i.e empty string code to the function here https://github.com/Expensify/App/blob/926d7af225d72feb3b8a958a1fc97fa99ee563fb/src/libs/ReportUtils.ts#L6379.

we can make https://github.com/Expensify/App/blob/1332a1c4c323fec29f7200e5e8a1c63e68bebeb2/src/libs/CurrencyUtils.ts#L100 this function here to make currency optional and leave as it is without sending any empty value this way we're ensured we never send any empty currency code to the function.

And then in the function instead of directly using the passed code we can use the `Intl.supportedValuesOf('currency')` to get the list of browser supported currencies and if it's not found we can reveret back to `USD` which will solve the issue.




### What alternative solutions did you explore? (Optional)

Sometimes we might get empty string (maybe from backend) so at that time we cannot do anything to face it, as proposed we can make the `currency` paramters optional here and instead of empty string we can pass `undefined` instead of empty string which will get replaced by default value proposed by the function defination.

<img width="611" alt="Screenshot 2024-06-03 at 11 33 46 PM" src="https://github.com/Expensify/App/assets/59088937/2b6eb3ee-1bdf-4630-ad32-cfa43bb077db">

**Comment by mountiny** [comment]
I think we should be ok to default to USD in this case as it should be very rare and its probably some malformed data

**Comment by mountiny** [comment]
Btw, I think the main problem here is that the `currency` is missing as a prop in the report object so its probably not correctly defaulting to USD

this is example of the report that crashes for me, no currency 
<img width="197" alt="image" src="https://github.com/Expensify/App/assets/36083550/e7ff5982-755b-48e1-90c3-76343ddb6a78">

**Comment by twisterdotcom** [comment]
Damn this is killing me today. Would love to get it fixed. Having to right click > Mark as read _a lot_

**Comment by ahmedGaber93** [comment]
Reviewing now

**Comment by AndrewGable** [comment]
I have this problem too

**Comment by ahmedGaber93** [comment]
Thanks all for the proposals.

I think we can move forward with @dragnoir's [proposal](https://github.com/Expensify/App/issues/43004#issuecomment-2145766525). Which remove fallback currency to empty string. This will let `convertToDisplayString` to use default value USD.

🎀 👀 🎀 C+ reviewed

**Comment by b4s36t4** [comment]
@ahmedGaber93 I'm sorry but I was the first one to propose that solution, can you re-verify it again?

**Comment by b4s36t4** [comment]
<img width="828" alt="Screenshot 2024-06-04 at 10 19 17 PM" src="https://github.com/Expensify/App/assets/59088937/e3b3b1d7-e2fa-4d1d-98bd-920d385ab939">

**Comment by b4s36t4** [comment]
Also I was the one to find the real root cause, all of the proposals are just tried to return false or return from Utils function, none of the proposals tried to explain the issue at first all.

**Comment by b4s36t4** [comment]
I don't think you read my proposal correctly, I did included code and which lines to edit (only example).

Also, no one found the root cause at all, which is causing the application to break. You all tried to raise the proposal first and then edited later on, and you did updated your proposal only after @mountiny said it's happening with empty string, I explained the issue before hand.

<img width="864" alt="Screenshot 2024-06-04 at 10 24 54 PM" src="https://github.com/Expensify/App/assets/59088937/175e5b6a-b820-4600-8b06-90257df2a7d7">

**Comment by ahmedGaber93** [comment]
> @ahmedGaber93 I'm sorry but I was the first one to propose that solution, can you re-verify it again?

@b4s36t4 Let me check that.

**Comment by ahmedGaber93** [comment]
> @ahmedGaber93, we should check for valid currency in convertToDisplayString function instead of fixing the use of it at every place.

I think this unnecessary step because the only case we faced is missing `currency` in the report, not unsupported currency.

**Comment by ahmedGaber93** [comment]
> @ahmedGaber93 I'm sorry but I was the first one to propose that solution, can you re-verify it again?

Yes, you are right, I missed check the edit history. Edit my review here https://github.com/Expensify/App/issues/43004#issuecomment-2147982864, also we should edit all places to ensure we never send empty string as currency.

**Comment by b4s36t4** [comment]
Thanks a lot @ahmedGaber93!

**Comment by ahmedGaber93** [comment]
@ShridharGoel I think check `Intl.NumberFormat` support is an unnecessary step, as we don't have a problem with what `Intl.NumberFormat` before, we just need to confirm we never pass `currency` to it as empty string.

**Comment by ahmedGaber93** [comment]
> In the future, we might still use CurrencyUtils.convertToDisplayString with an empty string, even if we fix all the current instances. This solution only resolves the current issue and is not future-proof.

I think for that, we can edit `currency` typescript to be `keyof CurrencyList | undefined` instead of string

**Comment by b4s36t4** [comment]
@dragnoir my proposal contains two changes (3 if we include the alternative as well) and one of the change is what you said, but pointing out this is the only solution I proposed seems a bit wrong.

Please read the proposal fully again!

**Comment by ahmedGaber93** [comment]
> But doesn't Intl.NumberFormat check ensure that the issue is handled well and is future-proof also since that's the place where the exception gets thrown?

@ShridharGoel Maybe this a good improvement, but the selected proposal has better check for that using `Intl.supportedValuesOf('currency')`

**Comment by srikarparsi** [comment]
I agree, let's go with @b4s36t4's proposal. Since the currency field is optional, we shouldn't be passing an empty string to it. And we can prevent empty strings in the future with the type definition as @ahmedGaber93 suggested.

**Comment by mallenexpensify** [comment]
@b4s36t4 please treat this issue with urgency because it's a reliability/quality bug.  Thx.

**Comment by CortneyOfstad** [comment]
PR has been created and QA is on-going 👍

**Comment by CortneyOfstad** [comment]
@ahmedGaber93 can you complete the checklist before EOD today so we're not delayed with payment tomorrow? Thanks!

**Comment by b4s36t4** [comment]
@CortneyOfstad doesn't the payment due date is yesterday which is the first that the resolution is deployed on?

**Comment by b4s36t4** [comment]
Also about the amount, shouldn't this come under critical issues which are eligible for 500$? I'm sorry if it's not!

**Comment by ahmedGaber93** [comment]
**BugZero Checklist**: The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [x] [@ahmedGaber93] The PR that introduced the bug has been identified. Link to the PR: N/A. the main issue is in backend.
- [x] [@ahmedGaber93] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: N/A
- [x] [@ahmedGaber93] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: N/A
- [x] [@ahmedGaber93] Determine if we should create a regression test for this bug. Yes
- [x] [@ahmedGaber93] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

### Regression Test Proposal


1. Open any report
2. Click (+) icon > Submit expense then submit an expense
3. Open the expense report that you created it in step 2 and Verify the app is not crashed

extra steps for Web and Desktop only

4. Open your browser console and type Onyx.merge("report_<your_report_id>",{currency: null}) then press Enter. You can get your_report_id from browser link.
5. Verify the app is not crashed

Do we agree 👍 or 👎

**Comment by CortneyOfstad** [comment]
## Payment Summary

@ahmedGaber93 — paid $500 (Critical) via Upwork
@b4s36t4 — paid $500 (Critical) via Upwork

## Regression

https://github.com/Expensify/Expensify/issues/406381

**Comment by mallenexpensify** [comment]
Heads up, in case anyone uses this as a reference for prices increases for 'critical' type issues.  [From #expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q/p1717451725980889?thread_ts=1712256409.152489&cid=C01GTK53T8Q)
> There have been questions about what constitutes a job being auto-increased to $500. These jobs are only the ones that have the High Priority label added to them (ie. issue with HIGH or CRITICAL in the title or project status/priority are not auto-increased to $500).
